dkg/dkg_test.go

```diff
@@ -292,53 +291,71 @@ func TestSyncFlow(t *testing.T) {
 			dkgs := make([]*testDkg, test.nodes)
 
 			var (
-				mu   sync.Mutex
-				wg   sync.WaitGroup
-				once bool                     // 'once' all initial peers are connected.
-				done = make(map[peer.ID]bool) // To not execute callback more than once per peer.
+				done           = make(chan struct{})
+				dkgErrChan     = make(chan error)
+				connectedCount int
 			)
-			callback := func(connected int, id peer.ID) {
-				mu.Lock()
-				defer mu.Unlock()
 
-				if once || done[id] || connected != len(test.connect)-1 {
-					return
-				}
+			newCallback := func(required int) func(int, peer.ID) {
+				var called bool
+				return func(connected int, id peer.ID) {
+					if called || connected != required {
+						return
+					}
 
-				done[id] = true
-				wg.Done()
+					called = true
+					done <- struct{}{}
+				}
 			}
 
 			// Start DKG for initial peers.
 			for _, idx := range test.connect {
-				wg.Add(1)
-				configs[idx].TestSyncCallback = callback
-				dkgs[idx] = startNewDKG(t, configs[idx], errChan)
+				configs[idx].TestSyncCallback = newCallback(len(test.connect) - 1)
+				dkgs[idx] = startNewDKG(t, ctx, configs[idx], dkgErrChan)
 			}
 
 			// Wait for initial peers to connect with each other.
-			wg.Wait()
+			for connectedCount != len(test.connect) {
+				select {
+				case <-done:
+					connectedCount++
+				case err = <-errChan:
+					cancel()
+					testutil.SkipIfBindErr(t, err)
+					require.Fail(t, fmt.Sprintf("bootnode error: %v", err))
+				case err = <-dkgErrChan:
+					cancel()
+					testutil.SkipIfBindErr(t, err)
+					require.Fail(t, fmt.Sprintf("dkg error: %v", err))
+				}
+			}
 
 			// Drop some peers.
 			for _, idx := range test.disconnect {
 				dkgs[idx].Stop()
 
 				// Wait for this dkg process to return.
-				<-dkgs[idx].runChan
+				err = <-dkgErrChan
+				require.ErrorContains(t, err, "context canceled")
 			}
 
-			// Do not execute callback again.
-			once = true
-
 			// Start remaining peers.
 			for _, idx := range test.reconnect {
-				dkgs[idx] = startNewDKG(t, configs[idx], errChan)
+				dkgs[idx] = startNewDKG(t, ctx, configs[idx], dkgErrChan)
 			}
 
 			// Assert DKG results
-			for i := 0; i < test.nodes; i++ {
-				dkgs[i].AssertDone(t)
+			select {
+			case err = <-errChan:
+				cancel()
+				testutil.SkipIfBindErr(t, err)
+				require.Fail(t, fmt.Sprintf("bootnode error: %v", err))
+			case err = <-dkgErrChan:
+				cancel()
```

Comment:
this cancels the parent context as soon as the first DKG completes, this might cause other DKGs to exit early and not write their output...

Reply:
addressed